core/include/gnuradio-4.0/Scheduler.hpp

```diff
@@ -171,6 +171,9 @@ class SchedulerBase {
                 if (progress_count == progress_count_old && done < n_batches) {
```

Comment:
```suggestion
                if (progress_count == progress_count_old && done + 1 < n_batches) {
```

I think this is the root problem, since the done count is only incremented in the `progress_new` and not in the underlying `done` in line 165. This should also be a problem for more than one thread.

Reply:
Thanks!  This is also valid change!
 If nothing happened, and progress remains the same and all threads are done (`done + 1 < n_batches`) no need to wait.

`done < n_batches` means that the last thread can still wait for the progress. It is strange why it did not show up in the test.